src/services/exit_order.py

```diff
@@ -1,24 +1,335 @@
-from src.web3py.typings import Web3
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Iterable
+
+from eth_typing import Address
+from web3.types import EventData
+
+from src.constants import FAR_FUTURE_EPOCH
+from src.modules.submodules.typings import ChainConfig
+from src.providers.consensus.typings import Validator
 
 from src.typings import BlockStamp
+from src.web3py.extentions.lido_validators import LidoValidator, NodeOperator, NodeOperatorIndex, StakingModuleId, \
+    NodeOperatorId
+from src.web3py.typings import Web3
+
+
+@dataclass
+class NodeOperatorPredictableState:
+    predictable_validators_total_age: int
+    predictable_validators_count: int
+    targeted_validators: int | None
+    delayed_validators: int
 
 
 class ValidatorsExit:
     """
-    Exit order is:
-    1. if NO active_keys < NO current keys
-    2. Stacking > 1% remove by sum validators age
-    3. By active validators numbers
+    Exit order predicates sequence:
+    1. Validator whose operator with the lowest number of delayed validators
+    2. Validator whose operator with the highest number of targeted validators to exit
+    3. Validator whose operator with the highest stake weight
+    4. Validator whose operator with the highest number of validators
+    5. Validator with the lowest activation epoch
+    6. Validator with the lowest index
     """
-    def __init__(self, blockstamp: BlockStamp, w3: Web3,  max_size: int):
-        self.blockstamp = blockstamp
+    staking_module_id: dict[Address, StakingModuleId]
+    lido_node_operator_stats: dict[NodeOperatorIndex, NodeOperatorPredictableState]
+    total_predictable_validators_count: int
+
+    def __init__(
+        self,
+        w3: Web3,
+        blockstamp: BlockStamp,
+        c_conf: ChainConfig,
+        exitable_lido_validators: list[LidoValidator],
+        max_validators_to_exit: int,
+    ):
         self.w3 = w3
+        self.blockstamp = blockstamp
```

Comment:
Here you save blockstamp to object attributes, but you are still providing blockstamp to methods in args

Reply:
fixed